### DISCUSSION TITLE
Adding my own libraries to the firmware

### DISCUSSION THREAD

**Discussion by liquikitten** [micropython/micropython#16118] [ESP32]
Hello, I encountered a problem combining the firmware with my .bin file. My .bin file contains .mpy scripts (made using mklittlefs). I was guided to combine this by this site https://docs.espressif.com/projects/esptool/en/latest/esp32s3/esptool/basic-commands.html
The problem is that when flashing the merged firmware, I cannot use my libraries, because for some reason they are not displayed in the list of available libraries. How can I properly add my libraries to micropython firmware?

**Comment by peterhinch** [comment]
The usual way of embedding modules in a firmware build is [documented here](https://docs.micropython.org/en/latest/reference/manifest.html).

**Reply by liquikitten** [reply]
I tried this method, but I got the following error when assembling
![2024-10-30 153101](https://github.com/user-attachments/assets/46f9302d-31cb-4b77-97ad-0a191019c919)

**Comment by robert-hh** [comment]
You have to set up the build environment for ESP32 properly as detailed in esp32/README.md. Note that you have to update  the submodules in the esp-idf directory after checking out the esp-idf version and to run the install script.  That's all documented in README.md. When that's done you can change back to the micropython/esp32 directory.
idf.py is located in <esp-idf_dir>/tools. Make sure that this is in your PATH. To set the proper paths, I run `. <idf_dir>/export.sh`.
